CUDA, ROCm and Alpaka-related updates

fwyzard opened the pull request:
#### PR description:

CUDA, ROCm and Alpaka-related updates:
  - add `PortableTestObjects` data formats for the ROCm Alpaka backend
  - add support for AMD GPUs via the `ProcessAcceleratorROCm`
  - update the description of the "backend" parameter
  - introduce different exit status codes for `cudaIsEnabled` and `rocmIsEnabled`
  - add abstract interfaces for the CUDA and ROCm services
  - do not load the CUDA or ROCm services if the backends are not available
  - update the `PhysicsTools/TensorFlow` tests accordingly

#### PR validation:

All unit tests should pass.

#### Backport status

To be backported to CMSSW 13.0.x for data taking.

fwyzard committed:
Add data formats for the ROCm Alpaka backend

fwyzard committed:
Add support for AMD GPUs via the ProcessAcceleratorROCm

fwyzard committed:
Update the "backend" parameter description

fwyzard committed:
Use different exit codes for different conditions

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40832/34287

- This PR adds an extra 32KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/StandardSequences/python/Accelerators_cff.py  modified in PR(s): #40725
  - File  DataFormats/PortableTestObjects/src/alpaka/classes_rocm.h  modified in PR(s): #40725
  - File  DataFormats/PortableTestObjects/src/alpaka/classes_rocm_def.xml  modified in PR(s): #40725
  - File  HeterogeneousCore/AlpakaCore/python/ProcessAcceleratorAlpaka.py  modified in PR(s): #40725
  - File  HeterogeneousCore/AlpakaCore/src/module_backend_config.cc  modified in PR(s): #40725
  - File  HeterogeneousCore/CUDACore/python/ProcessAcceleratorCUDA.py  modified in PR(s): #40725
  - File  HeterogeneousCore/CUDAServices/interface/CUDAService.h  modified in PR(s): #40725
  - File  HeterogeneousCore/CUDAServices/plugins/BuildFile.xml  modified in PR(s): #40725
  - File  HeterogeneousCore/CUDAServices/plugins/CUDAServiceDisabled.cc  modified in PR(s): #40725
  - File  HeterogeneousCore/CUDAServices/plugins/CUDAServiceEnabled.cc  modified in PR(s): #40725
  - File  HeterogeneousCore/CUDAServices/src/CUDAService.cc  modified in PR(s): #40725, #37952
  - File  HeterogeneousCore/CUDAServices/plugins/plugins.cc  modified in PR(s): #40725
  - File  HeterogeneousCore/CUDAServices/test/BuildFile.xml  modified in PR(s): #40725
  - File  HeterogeneousCore/CUDAServices/test/testCUDAService.cpp  modified in PR(s): #40725
  - File  HeterogeneousCore/CUDAServices/test/testCUDAService.py  modified in PR(s): #40725
  - File  HeterogeneousCore/CUDAServices/test/testCUDAServiceDisabled.cpp  modified in PR(s): #40725
  - File  HeterogeneousCore/CUDAServices/test/testCUDAServiceDisabled.py  modified in PR(s): #40725
  - File  HeterogeneousCore/CUDAServices/test/test_main.cpp  modified in PR(s): #40725
  - File  HeterogeneousCore/CUDAUtilities/bin/cudaIsEnabled.cpp  modified in PR(s): #40725
  - File  HeterogeneousCore/CUDAUtilities/bin/cudaIsEnabled_fallback.cpp  modified in PR(s): #40725
  - File  HeterogeneousCore/CUDAUtilities/interface/EventCache.h  modified in PR(s): #40725
  - File  HeterogeneousCore/CUDAUtilities/interface/StreamCache.h  modified in PR(s): #40725
  - File  HeterogeneousCore/Common/interface/PlatformStatus.h  modified in PR(s): #40725
  - File  HeterogeneousCore/Common/python/PlatformStatus.py  modified in PR(s): #40725
  - File  HeterogeneousCore/ROCmCore/python/ProcessAcceleratorROCm.py  modified in PR(s): #40725
  - File  HeterogeneousCore/ROCmCore/python/ProcessAcceleratorROCm_cfi.py  modified in PR(s): #40725
  - File  HeterogeneousCore/ROCmUtilities/bin/rocmIsEnabled.cpp  modified in PR(s): #40725
  - File  HeterogeneousCore/ROCmUtilities/bin/rocmIsEnabled_fallback.cpp  modified in PR(s): #40725

cmsbuild commented:
A new Pull Request was created by  @fwyzard (Andrea Bocci) for master.

It involves the following packages:

- Configuration/StandardSequences (**operations**)
- DataFormats/PortableTestObjects (**heterogeneous**)
- HeterogeneousCore/AlpakaCore (**heterogeneous**)
- HeterogeneousCore/CUDACore (**heterogeneous**)
- HeterogeneousCore/CUDAServices (**heterogeneous**)
- HeterogeneousCore/CUDAUtilities (**heterogeneous**)
- HeterogeneousCore/Common (**heterogeneous**)
- HeterogeneousCore/ROCmCore (**heterogeneous**)
- HeterogeneousCore/ROCmUtilities (**heterogeneous**)
- PhysicsTools/TensorFlow (**reconstruction**)


@perrotta, @rappoccio, @makortel, @clacaputo, @fwyzard, @mandrenguyen, @fabiocos, @davidlange6 can you please review it and eventually sign? Thanks.
@fabiocos, @VourMa, @makortel, @felicepantaleo, @riga, @GiacomoSguazzoni, @JanFSchulte, @rovere, @VinInn, @Martin-Grunewald, @missirol, @mmusich, @sameasy, @mtosi, @dgulhan, @slomeo this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals RelVals-INPUT AddOn 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/summary.html
**COMMIT**: 3ac12962547023b02f0cc77955c5159b988e34b3
**CMSSW**: CMSSW_13_1_X_2023-02-20-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30774/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @makortel cms-sw/cmssw#40827
 - @covarell cms-sw/cmssw#40824
 - @dinyar cms-sw/cmssw#40822
 - @kakwok cms-sw/cmssw#40829
 - @makortel cms-sw/cmssw#40792
 - @mmusich cms-sw/cmssw#40804

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/git-merge-result


## Unit Tests

I found errors in the following unit tests:

<pre>---> test testHeterogeneousCoreCUDATestWriteRead had ERRORS
---> test testHeterogeneousCoreAlpakaTestModules had ERRORS
---> test testHeterogeneousCoreAlpakaTestWriteRead had ERRORS
---> test TestCUDATest had ERRORS
and more ...
</pre>

## RelVals

- [139.001](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrix-results/139.001_RunMinimumBias2021)
```
----- Begin Fatal Exception 21-Feb-2023 11:04:07 CET-----------------------
An exception of category 'PluginLibraryLoadError' occurred while
   [0] Constructing the EventProcessor
   [1] While attempting to load plugin CUDAService
Exception Message:
unable to load /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30774/CMSSW_13_1_X_2023-02-20-2300/lib/el8_amd64_gcc11/poisoned/plugin-poisoned-HeterogeneousCoreCUDAServicesPlugins.so because /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30774/CMSSW_13_1_X_2023-02-20-2300/lib/el8_amd64_gcc11/poisoned/plugin-poisoned-HeterogeneousCoreCUDAServicesPlugins.so: cannot open shared object file: No such file or directory
----- End Fatal Exception -------------------------------------------------
```
- [11634.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrix-results/11634.0_TTbar_14TeV+2021)
```
----- Begin Fatal Exception 21-Feb-2023 11:17:29 CET-----------------------
An exception of category 'PluginLibraryLoadError' occurred while
   [0] Constructing the EventProcessor
   [1] While attempting to load plugin CUDAService
Exception Message:
unable to load /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30774/CMSSW_13_1_X_2023-02-20-2300/lib/el8_amd64_gcc11/poisoned/plugin-poisoned-HeterogeneousCoreCUDAServicesPlugins.so because /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30774/CMSSW_13_1_X_2023-02-20-2300/lib/el8_amd64_gcc11/poisoned/plugin-poisoned-HeterogeneousCoreCUDAServicesPlugins.so: cannot open shared object file: No such file or directory
----- End Fatal Exception -------------------------------------------------
```
- [11634.7](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrix-results/11634.7_TTbar_14TeV+2021_trackingMkFit)
```
----- Begin Fatal Exception 21-Feb-2023 11:17:29 CET-----------------------
An exception of category 'PluginLibraryLoadError' occurred while
   [0] Constructing the EventProcessor
   [1] While attempting to load plugin CUDAService
Exception Message:
unable to load /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30774/CMSSW_13_1_X_2023-02-20-2300/lib/el8_amd64_gcc11/poisoned/plugin-poisoned-HeterogeneousCoreCUDAServicesPlugins.so because /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30774/CMSSW_13_1_X_2023-02-20-2300/lib/el8_amd64_gcc11/poisoned/plugin-poisoned-HeterogeneousCoreCUDAServicesPlugins.so: cannot open shared object file: No such file or directory
----- End Fatal Exception -------------------------------------------------
```
<details>
<summary>Expand to see more relval errors ...</summary>

- [11634.911](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrix-results/11634.911_TTbar_14TeV+2021_DD4hep)
- [11634.914](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrix-results/11634.914_TTbar_14TeV+2021_DDDDB)
- [11834.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrix-results/11834.0_TTbar_14TeV+2021PU)
- [12434.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrix-results/12434.0_TTbar_14TeV+2023)
</details>



## RelVals-INPUT

- [139.002](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/139.002_RunZeroBias2021)```139.002_RunZeroBias2021/step2_RunZeroBias2021.log```
- [139.001](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/139.001_RunMinimumBias2021)```139.001_RunMinimumBias2021/step2_RunMinimumBias2021.log```
- [139.004](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/139.004_RunNoBPTX2021)```139.004_RunNoBPTX2021/step2_RunNoBPTX2021.log```
<details>
<summary>Expand to see more relval errors ...</summary>

- [139.003](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/139.003_RunHLTPhy2021)
- [140.001](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.001_RunMinimumBias2022A)
- [140.002](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.002_RunSingleMuon2022A)
- [140.003](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.003_RunZeroBias2022A)
- [140.005](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.005_RunJetHT2022A)
- [140.004](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.004_RunBTagMu2022A)
- [140.007](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.007_RunMET2022A)
- [140.006](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.006_RunDisplacedJet2022A)
- [140.008](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.008_RunEGamma2022A)
- [140.009](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.009_RunTau2022A)
- [140.01](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.01_RunDoubleMuon2022A)
- [140.021](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.021_RunMinimumBias2022B)
- [140.011](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.011_RunMuonEG2022A)
- [140.022](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.022_RunSingleMuon2022B)
- [140.023](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.023_RunZeroBias2022B)
- [140.024](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.024_RunBTagMu2022B)
- [140.025](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.025_RunJetHT2022B)
- [140.026](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.026_RunDisplacedJet2022B)
- [140.027](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.027_RunMET2022B)
- [140.029](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.029_RunTau2022B)
- [140.028](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.028_RunEGamma2022B)
- [140.03](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.03_RunDoubleMuon2022B)
- [140.031](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.031_RunMuonEG2022B)
- [140.043](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.043_RunZeroBias2022C)
- [140.042](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.042_RunSingleMuon2022C)
- [140.044](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.044_RunBTagMu2022C)
- [140.046](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.046_RunDisplacedJet2022C)
- [140.047](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.047_RunMET2022C)
- [140.045](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.045_RunJetHT2022C)
- [140.048](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.048_RunEGamma2022C)
- [140.049](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.049_RunTau2022C)
- [140.051](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.051_RunMuonEG2022C)
- [140.05](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.05_RunDoubleMuon2022C)
- [140.062](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.062_RunMuon2022D)
- [140.063](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.063_RunZeroBias2022D)
- [140.064](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.064_RunBTagMu2022D)
- [140.065](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.065_RunJetMET2022D)
- [140.066](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.066_RunDisplacedJet2022D)
- [140.067](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.067_RunEGamma2022D)
- [140.069](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.069_RunMuonEG2022D)
- [140.068](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.068_RunTau2022D)
- [140.104](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.104_RunZeroBias2022B)
- [140.105](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.105_RunBTagMu2022B)
- [140.106](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.106_RunJetHT2022B)
- [140.107](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.107_RunDisplacedJet2022B)
- [140.109](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.109_RunEGamma2022B)
- [140.111](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.111_RunDoubleMuon2022B)
- [140.11](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.11_RunTau2022B)
- [140.112](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.112_RunMuonEG2022B)
- [140.114](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.114_RunHcalNZS2022B)
- [140.113](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.113_RunNoBPTX2022B)
- [140.115](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.115_RunHLTPhysics2022B)
- [140.117](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.117_RunCosmics2022B)
- [140.116](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/140.116_RunCommissioning2022B)
- [11601.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11601.0_SingleElectronPt10+2021)
- [11602.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11602.0_SingleElectronPt35+2021)
- [11603.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11603.0_SingleElectronPt1000+2021)
- [11604.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11604.0_SingleGammaPt10+2021)
- [11605.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11605.0_SingleGammaPt35+2021)
- [11606.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11606.0_SingleMuPt1+2021)
- [11607.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11607.0_SingleMuPt10+2021)
- [11608.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11608.0_SingleMuPt100+2021)
- [11609.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11609.0_SingleMuPt1000+2021)
- [11634.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11634.0_TTbar_14TeV+2021)
- [11630.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11630.0_QCD_Pt_3000_3500_14TeV+2021)
- [11634.1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11634.1_TTbar_14TeV+2021_trackingOnly)
- [11634.15](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11634.15_TTbar_14TeV+2021_JMENano)
- [11634.24](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11634.24_TTbar_14TeV+2021_0T)
- [11634.5](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11634.5_TTbar_14TeV+2021_pixelTrackingOnly)
- [11634.501](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11634.501_TTbar_14TeV+2021_Patatrack_PixelOnlyCPU)
- [11634.505](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11634.505_TTbar_14TeV+2021_Patatrack_PixelOnlyTripletsCPU)
- [11634.511](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11634.511_TTbar_14TeV+2021_Patatrack_ECALOnlyCPU)
- [11634.521](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11634.521_TTbar_14TeV+2021_Patatrack_HCALOnlyCPU)
- [11634.7](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11634.7_TTbar_14TeV+2021_trackingMkFit)
- [11634.91](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11634.91_TTbar_14TeV+2021_trackdnn)
- [11640.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11640.0_MinBias_14TeV+2021)
- [11643.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11643.0_QCD_FlatPt_15_3000HS_14+2021)
- [11646.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11646.0_ZEE_14+2021)
- [11650.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11650.0_ZMM_14+2021)
- [11650.501](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11650.501_ZMM_14+2021_Patatrack_PixelOnlyCPU)
- [11650.505](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11650.505_ZMM_14+2021_Patatrack_PixelOnlyTripletsCPU)
- [11723.17](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11723.17_QCD_Pt_1800_2400_14+2021_seedingDeepCore)
- [11834.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11834.0_TTbar_14TeV+2021PU)
- [11834.13](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11834.13_TTbar_14TeV+2021PU_mlpf)
- [11834.19](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11834.19_TTbar_14TeV+2021PU_ecalDeepSC)
- [11834.24](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11834.24_TTbar_14TeV+2021PU_0T)
- [11834.21](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11834.21_TTbar_14TeV+2021PU_ProdLike)
- [11846.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11846.0_ZEE_14+2021PU)
- [12034.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/12034.0_TTbar_14TeV+2021Design)
- [11846.19](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/11846.19_ZEE_14+2021PU_ecalDeepSC)
- [12434.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/12434.0_TTbar_14TeV+2023)
- [12634.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/12634.0_TTbar_14TeV+2023PU)
- [12834.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/12834.0_TTbar_14TeV+2024)
- [13034.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/13034.0_TTbar_14TeV+2024PU)
- [20834.502](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/runTheMatrixINPUT-results/20834.502_TTbar_14TeV+2026D88_Patatrack_PixelOnlyGPU)
</details>



## AddOn Tests

- [hlt_mc_GRun](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/addOnTests/hlt_mc_GRun)
```
----- Begin Fatal Exception 21-Feb-2023 11:05:38 CET-----------------------
An exception of category 'PluginLibraryLoadError' occurred while
   [0] Constructing the EventProcessor
   [1] While attempting to load plugin CUDAService
Exception Message:
unable to load /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30774/CMSSW_13_1_X_2023-02-20-2300/lib/el8_amd64_gcc11/poisoned/plugin-poisoned-HeterogeneousCoreCUDAServicesPlugins.so because /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30774/CMSSW_13_1_X_2023-02-20-2300/lib/el8_amd64_gcc11/poisoned/plugin-poisoned-HeterogeneousCoreCUDAServicesPlugins.so: cannot open shared object file: No such file or directory
----- End Fatal Exception -------------------------------------------------
```
- [hlt_mc_GRun](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/addOnTests/hlt_mc_GRun)
```
----- Begin Fatal Exception 21-Feb-2023 11:06:26 CET-----------------------
An exception of category 'PluginLibraryLoadError' occurred while
   [0] Constructing the EventProcessor
   [1] While attempting to load plugin CUDAService
Exception Message:
unable to load /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30774/CMSSW_13_1_X_2023-02-20-2300/lib/el8_amd64_gcc11/poisoned/plugin-poisoned-HeterogeneousCoreCUDAServicesPlugins.so because /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30774/CMSSW_13_1_X_2023-02-20-2300/lib/el8_amd64_gcc11/poisoned/plugin-poisoned-HeterogeneousCoreCUDAServicesPlugins.so: cannot open shared object file: No such file or directory
----- End Fatal Exception -------------------------------------------------
```
- [hlt_mc_HIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/addOnTests/hlt_mc_HIon)
```
----- Begin Fatal Exception 21-Feb-2023 11:05:33 CET-----------------------
An exception of category 'PluginLibraryLoadError' occurred while
   [0] Constructing the EventProcessor
   [1] While attempting to load plugin CUDAService
Exception Message:
unable to load /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30774/CMSSW_13_1_X_2023-02-20-2300/lib/el8_amd64_gcc11/poisoned/plugin-poisoned-HeterogeneousCoreCUDAServicesPlugins.so because /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30774/CMSSW_13_1_X_2023-02-20-2300/lib/el8_amd64_gcc11/poisoned/plugin-poisoned-HeterogeneousCoreCUDAServicesPlugins.so: cannot open shared object file: No such file or directory
----- End Fatal Exception -------------------------------------------------
```
<details>
<summary>Expand to see more addon errors ...</summary>

- [hlt_mc_HIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/addOnTests/hlt_mc_HIon)
- [hlt_mc_PIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/addOnTests/hlt_mc_PIon)
- [hlt_mc_PIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/addOnTests/hlt_mc_PIon)
- [hlt_mc_PRef](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/addOnTests/hlt_mc_PRef)
- [hlt_mc_PRef](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/addOnTests/hlt_mc_PRef)
- [hlt_data_GRun](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/addOnTests/hlt_data_GRun)
- [hlt_data_GRun](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/addOnTests/hlt_data_GRun)
- [hlt_data_HIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/addOnTests/hlt_data_HIon)
- [hlt_data_HIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/addOnTests/hlt_data_HIon)
- [hlt_data_PIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/addOnTests/hlt_data_PIon)
- [hlt_data_PIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/addOnTests/hlt_data_PIon)
- [hlt_data_PRef](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/addOnTests/hlt_data_PRef)
- [hlt_data_PRef](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30774/addOnTests/hlt_data_PRef)
</details>

fwyzard committed:
CUDAService is no longer required by PixelCPEFast

fwyzard committed:
Drop the deviceWithMostFreeMemory() methods

fwyzard committed:
Clean up HcalRawToDigiGPU headers

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40832/34328

- This PR adds an extra 160KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDAServices/src/CUDAService.cc  modified in PR(s): #37952
  - File  RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorOnGPU.cc  modified in PR(s): #37952
  - File  RecoVertex/BeamSpotProducer/plugins/BeamSpotToCUDA.cc  modified in PR(s): #37952

cmsbuild commented:
Pull request #40832 was updated. @perrotta, @malbouis, @rappoccio, @yuanchao, @makortel, @ChrisMisan, @clacaputo, @tvami, @AdrianoDee, @srimanob, @saumyaphor4252, @fwyzard, @francescobrivio, @mandrenguyen, @fabiocos, @davidlange6 can you please check and sign again.

cmsbuild commented:
-1

**Failed Tests**:   Build  ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30843/summary.html
**COMMIT**: d9d14ff5c9a94f46bdb545c4397e6a1ced5c1327
**CMSSW**: CMSSW_13_1_X_2023-02-23-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30843/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>Copying tmp/el8_amd64_gcc11/src/PhysicsTools/TensorFlow/test/testTFConstSession/testTFConstSession to productstore area:
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-02-23-1100/src/PhysicsTools/TensorFlow/test/testConstSessionCUDA.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-02-23-1100/src/PhysicsTools/TensorFlow/test/testRunner.cpp
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-02-23-1100/src/PhysicsTools/TensorFlow/test/testBaseCUDA.h:26,
                 from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-02-23-1100/src/PhysicsTools/TensorFlow/test/testConstSessionCUDA.cc:13:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-02-23-1100/poison/HeterogeneousCore/CUDAServices/interface/CUDAService.h:1:2: error: #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
    1 | #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
      |  ^~~~~
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-02-23-1100/src/PhysicsTools/TensorFlow/test/testConstSessionCUDA.cc: In member function 'virtual void testConstSessionCUDA::test()':
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-02-23-1100/src/PhysicsTools/TensorFlow/test/testConstSessionCUDA.cc:49:16: error: 'CUDAInterface' was not declared in this scope
   49 |   edm::Service<CUDAInterface> cuda;
</pre>

## Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Entering Package RecoLocalCalo/HcalRecProducers
>> Entering Package RecoLocalTracker/SiPixelClusterizer
>> Entering Package RecoPixelVertexing/PixelTriplets
>> Entering Package RecoVertex/BeamSpotProducer
>> Compile sequence completed for CMSSW CMSSW_13_1_X_2023-02-23-1100
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 1
+ eval scram build outputlog '&&' '(python3' /data/cmsbld/jenkins/workspace/ib-run-pr-tests/cms-bot/buildLogAnalyzer.py --logDir /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-02-23-1100/tmp/el8_amd64_gcc11/cache/log/src '||' 'true)'
++ scram build outputlog
>> Entering Package Alignment/CommonAlignment
Entering library rule at src/Alignment/CommonAlignment/plugins
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-02-23-1100/src/Alignment/CommonAlignment/plugins/FilterOutLowPt.cc
</pre>

fwyzard committed:
Add abstract interfaces for the CUDA and ROCm services

Split the CUDA and ROCm services into an interface and a concrete
implementation.

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40832/34330

- This PR adds an extra 32KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDAServices/src/CUDAService.cc  modified in PR(s): #37952
  - File  RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorOnGPU.cc  modified in PR(s): #37952
  - File  RecoVertex/BeamSpotProducer/plugins/BeamSpotToCUDA.cc  modified in PR(s): #37952

cmsbuild commented:
Pull request #40832 was updated. @perrotta, @malbouis, @rappoccio, @yuanchao, @makortel, @ChrisMisan, @clacaputo, @tvami, @AdrianoDee, @srimanob, @saumyaphor4252, @fwyzard, @francescobrivio, @mandrenguyen, @fabiocos, @davidlange6 can you please check and sign again.

fwyzard committed:
Add abstract interfaces for the CUDA and ROCm services

Split the CUDA and ROCm services into an interface and a concrete
implementation.

Split the CUDAService into CUDAInterface and CUDAService.
The former implements the CUDAService interface, while the latter
implement the conrete functionality, and is used only if:
  - CUDA is available on the current platforms, OS and compiler;
  - the current system has at least one available CUDA GPU;
  - gpu-nvidia is among the process.options.accelerators.

Split the ROCmService into ROCmInterface and ROCmService.
The former implements the ROCmService interface, while the latter
implement the conrete functionality, and is used only if:
  - ROCm is available on the current platforms, OS and compiler;
  - the current system has at least one available ROCm GPU;
  - gpu-amd is among the process.options.accelerators.

Update all uses of the edm::Service<CUDAService> and edm::Service<ROCmService>
to the interface classes, and check explicitly that a concrete implementation
of the service is available.

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40832/34331

- This PR adds an extra 152KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDAServices/src/CUDAService.cc  modified in PR(s): #37952
  - File  RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorOnGPU.cc  modified in PR(s): #37952
  - File  RecoVertex/BeamSpotProducer/plugins/BeamSpotToCUDA.cc  modified in PR(s): #37952

cmsbuild commented:
Pull request #40832 was updated. @perrotta, @malbouis, @rappoccio, @yuanchao, @makortel, @ChrisMisan, @clacaputo, @tvami, @AdrianoDee, @srimanob, @saumyaphor4252, @fwyzard, @francescobrivio, @mandrenguyen, @fabiocos, @davidlange6 can you please check and sign again.

fwyzard commented:
please test for el8_aarch64_gcc11

fwyzard commented:
please test for el8_ppc64le_gcc11

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30852/summary.html
**COMMIT**: 769d230de7ed75af336af7a24314ab333efe8077
**CMSSW**: CMSSW_13_1_X_2023-02-22-2300/el8_ppc64le_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30852/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @fwyzard cms-sw/cmssw#40852
 - @Dr15Jones cms-sw/cmssw#40853

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30852/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30852/git-merge-result

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30849/summary.html
**COMMIT**: 769d230de7ed75af336af7a24314ab333efe8077
**CMSSW**: CMSSW_13_1_X_2023-02-23-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30849/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



There are some workflows for which there are errors in the baseline:
 22034.911 step 1
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

**Summary**:
 - You potentially removed 13 lines from the logs
 - Reco comparison results: 5 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3528709
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3528684
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-02-23-1100+9dbeaa/55689/triggerResults): no differences found

fwyzard commented:
+heterogeneous

fwyzard commented:
@cms-sw/alca-l2 this PR is purely technical; could you check it and sign it, along with its 13.0.x backport, #40725 ?

fwyzard commented:
@cms-sw/reconstruction-l2 this PR is purely technical; could you check it and sign it, along with its 13.0.x backport, #40725 ?

fwyzard commented:
@cms-sw/operations-l2 this PR is purely technical; could you check it and sign it, along with its 13.0.x backport, #40725 ?

fwyzard commented:
@cms-sw/upgrade-l2 this PR is purely technical; could you check it and sign it, along with its 13.0.x backport, #40725 ?

tvami commented:
+alca
 * technical, test pass, small change to AlCa pkgs

fwyzard commented on HeterogeneousCore/AlpakaCore/python/ProcessAcceleratorAlpaka.py:
```diff
@@ -45,26 +51,70 @@ class ProcessAcceleratorAlpaka(cms.ProcessAccelerator):
+            process.MessageLogger.AlpakaService = cms.untracked.PSet()
+
+        # Check if the CPU backend is available
+        try:
+            if not "cpu" in accelerators:
+                raise False
             from HeterogeneousCore.AlpakaServices.AlpakaServiceSerialSync_cfi import AlpakaServiceSerialSync
```
Yes - for example, at the moment `AlpakaServiceROCmAsync_cfi` does not exist, because ROCm is not enabled among the Alpaka back-ends.

fwyzard commented on HeterogeneousCore/AlpakaCore/python/ProcessAcceleratorAlpaka.py:
```diff
@@ -45,26 +51,70 @@ class ProcessAcceleratorAlpaka(cms.ProcessAccelerator):
+
+        # Check if CUDA is available, and if the system has at least one usable NVIDIA GPU
+        try:
+            if not "gpu-nvidia" in accelerators:
+                raise False
+            status = PlatformStatus(os.waitstatus_to_exitcode(os.system("cudaIsEnabled")))
+            if status != PlatformStatus.Success:
+                raise False
```
Ehm, maybe ?
I mean that I do not know how the `accelerators` list gets populated, and if by the time `ProcessAcceleratorAlpaka.apply()` is called `gpu-nvidia` would already have been removed if there is no viable NVIDIA GPU in the system.

makortel replied:
The `apply()` function's `accelerators` argument is guaranteed to contain only those accelerator labels that are selected (via `process.options.accelerators`) and available (as dictated by `enabledLabels()). The availability of all accelerators is resolved first before the `apply()` function of any `ProcessAccelerator` is called. Therefore `ProcessAcceleratorAlpaka.apply()` should not see `gpu-nvidia` in case there system does not have a viable NVIDIA GPU.

fwyzard commented on HeterogeneousCore/CUDACore/README.md:
```diff
@@ -83,7 +83,8 @@ This page documents the CUDA integration within CMSSW
 4. Outside of `acquire()`/`produce()`, CUDA API functions may be
-   called only if `CUDAService::enabled()` returns `true`.
+   called only if `CUDAService` is available and `CUDAService::enabled()`
+   returns `true`.
```
Do you mean something like
```suggestion
   returns `true`:
     ```c++
     edm::Service<CUDAInterface> cuda;
     if (cuda and cuda->enabled()) {
       // CUDA calls ca be made here
     }
     ```
```
?

makortel replied:
A code example is even better than what I had in mind.

fwyzard commented on HeterogeneousCore/CUDACore/python/ProcessAcceleratorCUDA.py:
```diff
@@ -2,29 +2,53 @@
-        if not hasattr(process, "CUDAService"):
-            from HeterogeneousCore.CUDAServices.CUDAService_cfi import CUDAService
-            process.add_(CUDAService)
+        self._check()
```
OK, so just to confirm, it is guaranteed that `apply()` is always called after `enabledLabels()`, and that both are called on the "worker node" ?

makortel replied:
> OK, so just to confirm, it is guaranteed that `apply()` is always called after `enabledLabels()`, and that both are called on the "worker node" ?

Yes, this is guaranteed.

fwyzard commented on HeterogeneousCore/CUDACore/python/ProcessAcceleratorCUDA.py:
```diff
@@ -2,29 +2,53 @@
-            process.add_(CUDAService)
+        self._check()
+
+        if self._status == PlatformStatus.Success and self._label in accelerators:
```
And, as a consequence, that `accelerators` contains `self._label` _if and only if_ it was returned by `enabledLabels()` ?

makortel replied:
> And, as a consequence, that `accelerators` contains `self._label` _if and only if_ it was returned by `enabledLabels()` ?

Correct.

makortel replied:
> > And, as a consequence, that `accelerators` contains `self._label` _if and only if_ it was returned by `enabledLabels()` ?
> 
> Correct.

Well, there is still the further requirement for `accelerators` to contain `self._label` that `process.options.accelerators` must contain a wildcard that matches `self._label`.

fwyzard replied:
Right.... then

> `accelerators` contains `self._label` _only if_ it was returned by `enabledLabels()`

fwyzard commented:
please test for el8_aarch64_gcc11

cmsbuild commented:
Pull request #40832 was updated. @perrotta, @malbouis, @rappoccio, @yuanchao, @makortel, @ChrisMisan, @clacaputo, @tvami, @AdrianoDee, @srimanob, @saumyaphor4252, @fwyzard, @francescobrivio, @mandrenguyen, @fabiocos, @davidlange6 can you please check and sign again.

fwyzard commented:
+heterogeneous

fwyzard commented:
please test for el8_aarch64_gcc11

fwyzard commented:
please test for el8_ppc64le_gcc11

fwyzard commented:
please test for el8_amd64_gcc11

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30907/summary.html
**COMMIT**: d2ba4608a0e32fd9f66ae0c5c468eee63a6847c0
**CMSSW**: CMSSW_13_1_X_2023-02-24-2300/el8_aarch64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30907/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test testCudaCheck had ERRORS
</pre>

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30908/summary.html
**COMMIT**: d2ba4608a0e32fd9f66ae0c5c468eee63a6847c0
**CMSSW**: CMSSW_13_1_X_2023-02-24-2300/el8_ppc64le_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40832/30908/install.sh` to create a dev area with all the needed externals and cmssw changes.

cmsbuild commented:
-1

**Failed Tests**: RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30910/summary.html
**COMMIT**: d2ba4608a0e32fd9f66ae0c5c468eee63a6847c0
**CMSSW**: CMSSW_13_1_X_2023-02-26-0000/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40832/30910/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-INPUT

- [1002.4](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30910/runTheMatrixINPUT-results/1002.4_RunDoubleMuon2022B)```1002.4_RunDoubleMuon2022B/step2_RunDoubleMuon2022B.log```
- [20834.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30910/runTheMatrixINPUT-results/20834.0_TTbar_14TeV+2026D88)```20834.0_TTbar_14TeV+2026D88/step2_TTbar_14TeV+2026D88.log```
- [20834.103](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30910/runTheMatrixINPUT-results/20834.103_TTbar_14TeV+2026D88Aging3000)```20834.103_TTbar_14TeV+2026D88Aging3000/step2_TTbar_14TeV+2026D88Aging3000.log```
<details>
<summary>Expand to see more relval errors ...</summary>

- [20834.21](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30910/runTheMatrixINPUT-results/20834.21_TTbar_14TeV+2026D88_ProdLike)
- [20834.5](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30910/runTheMatrixINPUT-results/20834.5_TTbar_14TeV+2026D88_pixelTrackingOnly)
- [20834.501](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30910/runTheMatrixINPUT-results/20834.501_TTbar_14TeV+2026D88_Patatrack_PixelOnlyCPU)
- [20834.502](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30910/runTheMatrixINPUT-results/20834.502_TTbar_14TeV+2026D88_Patatrack_PixelOnlyGPU)
- [20834.75](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30910/runTheMatrixINPUT-results/20834.75_TTbar_14TeV+2026D88_HLT75e33)
- [20834.76](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30910/runTheMatrixINPUT-results/20834.76_TTbar_14TeV+2026D88_HLTwDIGI75e33)
- [20834.9](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30910/runTheMatrixINPUT-results/20834.9_TTbar_14TeV+2026D88_vectorHits)
- [21034.21](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30910/runTheMatrixINPUT-results/21034.21_TTbar_14TeV+2026D88PU_ProdLike)
- [21034.114](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30910/runTheMatrixINPUT-results/21034.114_TTbar_14TeV+2026D88PU_OTInefficiency10PC)
</details>



## Comparison Summary



**Summary**:
 - You potentially removed 15 lines from the logs
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3528955
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3528927
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-02-26-0000+9dbeaa/55729/triggerResults): no differences found

fwyzard commented:
urgent

fwyzard commented:
this should be in pre1

fwyzard committed:
Sync with CMSSW_13_1_X

fwyzard commented:
Thanks to @missirol for suggesting [a fix](https://github.com/cms-sw/cmssw/compare/d2ba4608a0e32fd9f66ae0c5c468eee63a6847c0..adc5096b34c08ef2bf6116a2ba549f8755a9ccb6) to the BuildFiles.

fwyzard commented:
Rebased on top of the latest IB.

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40832/34375

- This PR adds an extra 172KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDAServices/src/CUDAService.cc  modified in PR(s): #37952
  - File  RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorOnGPU.cc  modified in PR(s): #37952
  - File  RecoVertex/BeamSpotProducer/plugins/BeamSpotToCUDA.cc  modified in PR(s): #37952

cmsbuild commented:
Pull request #40832 was updated. @perrotta, @malbouis, @rappoccio, @yuanchao, @makortel, @ChrisMisan, @clacaputo, @tvami, @AdrianoDee, @srimanob, @saumyaphor4252, @fwyzard, @francescobrivio, @mandrenguyen, @fabiocos, @davidlange6 can you please check and sign again.

fwyzard commented:
+heterogeneous

tvami commented:
+alca

mandrenguyen commented:
+reconstruction

smuzaffar commented:
@fwyzard , i see that this PR `classes h/def` files for `rocm`, so should this be tested with https://github.com/cms-sw/cmsdist/pull/8301 (where we enable `rocm` backend)?

fwyzard commented:
This PR can go in also without cms-sw/cmsdist#8301 - the ROCm part will just be ignored.

In cms-sw/cmsdist#8301 I asked the bot to run the tests with this PR, so we can se the effect of enabling ROCm.

cmsbuild commented:
-1

**Failed Tests**:   Build 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30949/summary.html
**COMMIT**: adc5096b34c08ef2bf6116a2ba549f8755a9ccb6
**CMSSW**: CMSSW_13_1_X_2023-02-27-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40832/30949/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-02-27-2300/src/PhysicsTools/TensorFlow/test/testRunner.cpp
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-02-27-2300/src/PhysicsTools/TensorFlow/test/testVisibleDevicesCUDA.cc
>> Building binary testTFVisibleDevicesCUDA
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02774/el8_amd64_gcc11/external/gcc/11.2.1-f9b9dfdd886f71cd63f5538223d8f161/bin/../lib/gcc/x86_64-redhat-linux-gnu/11.2.1/../../../../x86_64-redhat-linux-gnu/bin/ld: tmp/el8_amd64_gcc11/src/PhysicsTools/TensorFlow/test/testTFVisibleDevicesCUDA/ccgiXMWj.ltrans0.ltrans.o: in function `testVisibleDevicesCUDA::test()':
<artificial>:(.text+0x28e5): undefined reference to `cms::cudatest::testDevices()'
collect2: error: ld returned 1 exit status
>> Deleted: tmp/el8_amd64_gcc11/src/PhysicsTools/TensorFlow/test/testTFVisibleDevicesCUDA/testTFVisibleDevicesCUDA
gmake: *** [tmp/el8_amd64_gcc11/src/PhysicsTools/TensorFlow/test/testTFVisibleDevicesCUDA/testTFVisibleDevicesCUDA] Error 1
>> Leaving Package PhysicsTools/TensorFlow
>> Package PhysicsTools/TensorFlow built
>> Subsystem Utilities built
</pre>

fwyzard committed:
Fix dependencies in the test BuildFiles

fwyzard commented:
Sorry about pushing another update - but the previous change revealed some missing dependencies in the `PhysicsTools/TensorFlow/test` BuildFiles.

fwyzard commented:
+heterogeneous

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40832/34377

- This PR adds an extra 12KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDAServices/src/CUDAService.cc  modified in PR(s): #37952
  - File  RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorOnGPU.cc  modified in PR(s): #37952
  - File  RecoVertex/BeamSpotProducer/plugins/BeamSpotToCUDA.cc  modified in PR(s): #37952

cmsbuild commented:
Pull request #40832 was updated. @perrotta, @malbouis, @rappoccio, @yuanchao, @ChrisMisan, @clacaputo, @tvami, @AdrianoDee, @srimanob, @saumyaphor4252, @francescobrivio, @mandrenguyen, @fabiocos, @davidlange6 can you please check and sign again.

mandrenguyen commented:
+1
resign

tvami commented:
+alca

AdrianoDee commented:
+upgrade
(upgrade-wise only changes to few headers)

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9dbeaa/30954/summary.html
**COMMIT**: 1c30d1c009f353bd69a07fb6d0d517a0161616b5
**CMSSW**: CMSSW_13_1_X_2023-02-27-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40832/30954/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 29 lines from the logs
 - Reco comparison results: 9 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3529699
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3529671
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-02-27-2300+9dbeaa/55765/triggerResults): no differences found

rappoccio commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will be automatically merged.
